### PR TITLE
cli: esm support

### PR DESCRIPTION
**Pull request by davidjgoss**
## Motivation

#1304 

Support ESM usage with minimal additional flags/configuration and without dropping support for Node 10.

## Details

To allow importing cucumber functions from ESM support code, provide a "wrapper" mjs entry point for ESM projects as described in https://nodejs.org/api/packages.html#packages_approach_1_use_an_es_module_wrapper.

Usage:

```
cucumber-js --esm
```

The `--esm` flag causes us to use `import()` instead of `require()` for loading support code - the former works for both ESM and commonjs, where the latter only works for commonjs. When we drop Node 10 support, we'll be able to deprecate/remove the flag and just use `import()` across the board.

I think we need to consider usage of this in conjunction with on-the-fly transpilers as out of scope for now. For example ts-node is at a very experimental phase with ESM per https://github.com/TypeStrong/ts-node/issues/1007.

Test project: https://github.com/davidjgoss/cucumber-esm-example

### Todo

- [x] Use `import()` or `require()` appropriately at runtime for support code
- [x] Handle importing cucumber.js configuration file
- [x] Handle importing custom formatters
- [x] Handle importing snippets
- [x] Handle parallel runtime
- [x] Include `.mjs` files by default if using ESM
- [x] Test usage with ESM in CI
- [x] Resolve Windows issues
- [x] Documentation

### CODE REVIEW

**Comment by davidjgoss** [comment]
Able to do successful runs locally with some manual changes to compiled code - TypeScript transforms the dynamic imports down to requires, this might be because of `esModuleInterop`, need to investigate further as removing that option causes a lot of import errors.

Another option might be to have two compile targets, one for commonjs and one for ESM. This has some potential issues:

- We have internal state, so there's a risk with projects that could hit both sets of code. Given cucumber-js is often wrapped into other test frameworks, this seems like a possibility
- TypeScript doesn't allow us to specify the output file extension, so we can't have one set of files come out as `.cjs` and another as `.mjs` - even if we renamed as a build step, the require/import statements wouldn't be right

See also Node's documentation: https://nodejs.org/api/packages.html#packages_dual_commonjs_es_module_packages

Right now the most promising path looks like to create a separate `.mjs` entry point that re-exports the commonjs one. This gets away from the state issue, and may also allow us to set something that flips the support code loading from `require()` to `import()`.

**Comment by davidjgoss** [comment]
@charlierudolph some work to do still but would be good to get your take on this.

We're dealing with this on two fronts:

- An alternative executable `cucumber-es` runs cucumber with an override that loads support code via dynamic imports
- An wrapper entry point for ESM so that support code can successfully `import` cucumber functions

**Comment by charlierudolph** [comment]
Cool. I think approach makes sense to me.

To check my understanding we need a second binary as we need node to load things differently?

I would prefer if we could have a `--esm` flag on our main binary that then just calls to second binary so our users don't need to know about the second binary.

File: bin/cucumber-es.mjs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env node
```
Comment (charlierudolph) [inline bin/cucumber-es.mjs:1]: Think I would prefer this being called `cucumber-js-esm`

**Comment by davidjgoss** [comment]
@charlierudolph 

> I would prefer if we could have a `--esm` flag on our main binary

Done - was able to make this work by keeping an `importer` abstraction off the TypeScript compilation path, works nicely.

The only thing left now is Windows, where trying to `await import()` absolute file paths has issues - see https://github.com/nodejs/node/issues/31710.

File: src/formatter/builder.ts [hunk 2]
```diff
@@ -99,28 +110,43 @@ const FormatterBuilder = {
...
+      Syntax = FormatterBuilder.resolveConstructor(Syntax)
     }
     return new StepDefinitionSnippetBuilder({
       snippetSyntax: new Syntax(snippetInterface),
       parameterTypeRegistry: supportCodeLibrary.parameterTypeRegistry,
     })
   },
 
-  loadCustomFormatter(customFormatterPath: string, cwd: string) {
-    const CustomFormatter = createRequire(cwd)(customFormatterPath)
```
Comment (davidjgoss) [inline src/formatter/builder.ts:111]: Unfortunately had to add some fragmentation back here, as the ESM equivalent of `createRequire` is [not yet stable](https://nodejs.org/api/esm.html#esm_import_meta_resolve_specifier_parent).

File: src/formatter/builder.ts [hunk 3]
```diff
@@ -99,28 +110,43 @@ const FormatterBuilder = {
     let Syntax = JavascriptSnippetSyntax
     if (doesHaveValue(snippetSyntax)) {
       const fullSyntaxPath = path.resolve(cwd, snippetSyntax)
-      Syntax = require(fullSyntaxPath) // eslint-disable-line @typescript-eslint/no-var-requires
+      Syntax = await importer(fullSyntaxPath, true)
```
Comment (davidjgoss) [inline src/formatter/builder.ts:113]: Occurs to me that we should add the same logic as was done for formatters, to differentiate between project-relative paths and module names - but that would technically be a breaking change? Maybe in a future PR

Comment (charlierudolph) [reply src/formatter/builder.ts:113]: Created by #1640

File: bin/cucumber-js [hunk 4]
```diff
@@ -1,3 +1,3 @@
 #!/usr/bin/env node
 
-require('../lib/cli/run.js').default();
+require('../lib/cli/run.js').default()
```
Comment (charlierudolph) [inline bin/cucumber-js:3]: What's the reason for this change?

File: src/cli/configuration_builder_spec.ts [hunk 5]
```diff
@@ -65,52 +66,114 @@ describe('Configuration', () => {
...
+          featurePaths,
+          pickleFilterOptions,
+          supportCodePaths,
+        } = await ConfigurationBuilder.build({ argv, cwd })
+
+        // Assert
+        expect(featurePaths).to.eql([featurePath])
+        expect(pickleFilterOptions.featurePaths).to.eql([relativeFeaturePath])
+        expect(supportCodePaths).to.eql([jsSupportCodePath, mjsSupportCodePath])
+      })
```
Comment (charlierudolph) [inline src/cli/configuration_builder_spec.ts:176]: I don't think we need nested feature tests for non-esm. Thoughts on the following:
```js
describe('path to a feature', () => {
  describe('without esm', () => {})
  describe('with esm and js support files', () => {})
  describe('with esm and mjs support files', () => {})
});

describe('path to a nested feature', () => {}) // unchanged
```

File: bin/cucumber-js (hunk 4 above)
Comment (davidjgoss) [reply bin/cucumber-js:3]: Ah, I think this was just an artifact of me initially making some changes here and then backing them out. Reverted now for clarity.

File: src/cli/configuration_builder_spec.ts (hunk 5 above)
Comment (davidjgoss) [reply src/cli/configuration_builder_spec.ts]: Makes sense - updated per your suggestion.

File: docs/cli.md [hunk 6]
```diff
@@ -81,6 +81,14 @@ You can pass in format options with `--format-options <JSON>`. The JSON string m
 
 * Suggested use: add with profiles so you can define an object and use `JSON.stringify` instead of writing `JSON` manually.
 
+## ES Modules (experimental)
+
+You can optionally write your support code (steps, hooks, etc) with native ES modules syntax - i.e. using `import` and `export` statements without transpiling.
+
+To enable this, run with the `--esm` flag.
+
+This will also expand the default glob for support files to include the `.mjs` file extension.
```
Comment (charlierudolph) [inline docs/cli.md:90]: Is ESM purely additive (ie require is still supported)?

If not, would seem like using this requires custom formatters used would also need to be written in ESM

Comment (davidjgoss) [reply docs/cli.md:90]: ESM is additive. If enabled, it'll use `import()` to load all user code (steps/hooks, formatters, snippets, and config file). If any of this user is CommonJS and has `require()`s in it, that will still work fine - [you can `import()` CommonJS modules, but you can't `require()` ESM](https://nodejs.org/api/esm.html#esm_interoperability_with_commonjs).

(Hope I've understood the question right!)

Comment (davidjgoss) [reply docs/cli.md:90]: And I'll push some documentation improvements to this effect.

File: importers.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+/*
```
Comment (charlierudolph) [inline importers.js:1]: Thoughts on putting this in src and as part of building copy it into lib? Same for wrappers.mjs.

**Review by charlierudolph — APPROVED** [review]
Just the last comment about the location of a couple files. Otherwise looks great! Thanks for doing this!

File: docs/cli.md (hunk 6 above)
Comment (charlierudolph) [reply docs/cli.md:90]: Looks good. Thanks!

File: importers.js (hunk 7 above)
Comment (davidjgoss) [reply importers.js]: Good point - done!
